### PR TITLE
Rework of Weather Management

### PR DESCRIPTION
**Pull request by Petap0w**
__Complete rethink of Weather management including:__
A - Weather change alerts from Weather Webhooks (RDM included)
B - Weather change alerts from boosted Pokémons in the s2cell
C - Weather Forecast based on AccuWeather data to show on Pokémon alerts if the weather changes are likely to change boost status and stats
All those features are independent and can be set in config with the new `weather` section:
```
    "weather": {
        "weatherChangeAlert": false,                  // To enable or disable the weather change alerts
        "showAlteredPokemon": false,               // Allow weather changed pokémon in DTS
        "showAlteredPokemonStaticMap": false,   // Show weather changed on static map
        "showAlteredPokemonMaxCount": 10,      // Max number of changed pokémon allowed per alert
        "enableWeatherForecast": false,            // To enable or disable weather forecast
        "apiKeyAccuWeather": ["key1", "key2", "key3"],       // AccuWeather API keys
        "apiKeyDayQuota": 50,                           // Maximum API calls allowed per day
        "smartForecast": false,                          // use smart update of weather forecast
        "localFirstFetchHOD": 3,                        // First hour of the day for the first API call
        "forecastRefreshInterval": 8                  // API call interval
    },
```

__Requirements:__
- `npm install` in PoracleJS folder to install new dependencies (`nodes2ts`)
- Webhook of the `weather` type for feature A (`config > general > disableWeather` set to `false`)
- Just the normal `pokemon` type webhooks for feature B (`config > general > disablePokemon` set to `false`)
- AccuWeather API app key for feature C
Register on AccuWeather website and create an APP which will grant you an API key. You can make 50 calls per day with one key. Depending on the interval you'll set to fetch forecast data, this can cover around 15 level 10 s2cells per day (with getting weather every 8 hours)

__Features A and B:__
When weather in a s2cell changes compared to last hour, all the users (or channels) who were caring about a Pokémon in that cell will receive a Weather change alert:
Capture d’écran 2021-01-16 à 16 32 26
This example is using the new `pokemon-weather` tileserver template available in the example folder (copy to your tileserver). The template expects `{latitude: double, longitude: double, condition: int, coords: [[lat1,lon1],[lat2,lon2],[lat3,lon3],[lat4,lon4]]}` and is fully integrated with the `tileservercache` staticProvider option in Poracle.

If the config option `showAlteredPokemon` is set to true, each user will receive a custom alert showing the pokémons he cared for concerned by the weather change (with the limit `showAlteredPokemonMaxCount`).
Then if `showAlteredPokemonStaticMap` is set to true, the static map will show them as well.
Capture d’écran 2021-01-18 à 10 30 33

__Feature C:__
When a Pokémon webhook is analyzed, and if it despawns after the next hour mark, if the weather for next hour is different and will not boost anymore / will newly boost the Pokémon, then some new DTS are available on the alert:
```
weatherChange -> an already formed sentence for this weather change
weatherCurrent -> current weather ID
weatherCurrentName -> current weather
weatherCurrentEmoji -> current weather emoji
weatherNext -> weather ID forecast for next hour
weatherNextName -> weather name for next hour
weatherNextEmoji -> weather emoji for next hour
weatherChangeTime -> change time (format HH:00)
```
Giving:
Capture d’écran 2021-01-16 à 16 42 14
When is the weather forecast pulled?
* If `smartForecast` is set to true : when a pokémon is cared for in a s2cell and if no forecast is already available it is pulled for this cell, and a timeout of `forecastRefreshInterval` hours is set to refresh. Refresh will only happen if a pokémon is cared for at the timeout moment. Refresh will be forced in the cell if a pokémon is detected with a boost status not matching the forecast pulled for this hour (forced refresh will occur next time a pokémon is actually cared for in the cell).
* If `smartForecast` is set to false : when a pokémon is cared for in a s2cell and if no forecast is available it is pulled and a timeout is set at the next fixed hour defined by the combination of `localFirstFetchHOD` and `forecastRefreshInterval` (example with `localFirstFetchHOD = 3` and `forecastRefreshInterval = 8` at 3am, 11am and 7pm)

### CODE REVIEW

**Comment by ccev** [comment]
A possible feature you could add would be the editing of a message to change `Weather might change` to `Weather & Stats changed`/`Weather did not change`.
No idea about Telegram, but Discord Bots and [recently Webhooks](https://discord.com/developers/docs/resources/webhook#modify-webhook) would allow this.

**Comment by muckelba** [comment]
> No idea about Telegram, but Discord Bots and [recently Webhooks](https://discord.com/developers/docs/resources/webhook#modify-webhook) would allow this.

Telegram is able to do that. Should be pretty easy to implement because Poracle already have those message IDs to clean them later. Docs: https://core.telegram.org/bots/api#editmessagetext

**Comment by acocalypso** [comment]
Looks good for me, no issues found so far.

**Comment by kamieniarz** [comment]
I've been testing it for few days also and it works flawless

**Comment by t00fan** [comment]
Been testing this for the last several days, and it's been working great!

File: src/app.js [hunk 1]
```diff
@@ -318,17 +324,31 @@ async function processOne(hook) {
...
+				const hookHourTimestamp = updateTimestamp - (updateTimestamp % 3600)
+				if (fastify.cache.has(`${hook.message.s2_cell_id}_${hookHourTimestamp}`)) {
+					fastify.logger.debug(`Weather for ${hook.message.s2_cell_id} was sent again too soon, ignoring`)
+					break
+				}
+				fastify.cache.set(`${hook.message.s2_cell_id}_${hookHourTimestamp}`, 'cached')
+				const result = await fastify.weatherController.handle(hook.message)
+				result.forEach((job) => {
+					if (['discord:user', 'discord:channel', 'webhook'].includes(job.type)) fastify.discordQueue.push(job)
+					if (['telegram:user', 'telegram:channel'].includes(job.type)) fastify.telegramQueue.push(job)
```
Comment (jfberry) [inline src/app.js:350]: think telegram:group missing here

**Comment by kbtbc** [comment]
Been running this for a couple of days (only change notifications, no forecasting)...   seems to be working well.   Some users like them, some are asking for a way to disable them.  I don't know how complicated it would be, but a user-level toggle to choose to get weather notifications or not has been requested.

File: config/defaults/dts.json [hunk 2]
```diff
@@ -182,8 +182,11 @@
     "platform": "discord",
     "template": {
       "embed": {
-        "title": "Weather change!",
-        "description": "NB! The weather for some active mons have changed from {{oldweather}} {{oldweatheremoji}} to {{weather}} {{weatheremoji}}!\nThis could have altered the reported IV!"
+        "title": "⚠️ Weather change ⚠️",
+        "description": "The weather for some active Pokémons you are tracking has changed!\n{{#if oldweather}}It went from {{oldweather}} {{oldweatheremoji}} to {{else}}It is now {{/if}}{{weather}} {{weatheremoji}}\n{{#if activePokemons}}The following Pokémons have now changed and will have different stats:\n{{#each activePokemons}}**{{this.name}}** {{#isnt this.formname 'Normal'}} {{this.formname}}{{/isnt}} - {{round this.iv}}% - {{this.cp}}CP\n{{/each}}{{else}}This could have altered the reported stats and IV...{{/if}}",
```
Comment (kbtbc) [inline config/defaults/dts.json:186]: The plural here should be 'Pokémon'.   Line 199, too

**Review by kbtbc — CHANGES_REQUESTED** [review]

**Review by kbtbc — APPROVED** [review]
Approved!  🥇  lol

**Pull request merged**